data/transactions/verify/txn.go

```diff
@@ -208,18 +204,36 @@ func stxnVerifyCore(s *transactions.SignedTxn, txnIdx int, groupCtx *GroupContex
 		numSigs++
 		hasLogicSig = true
 	}
-	if numSigs == 0 {
-		// Special case: special sender address can issue special transaction
-		// types (compact cert txn) without any signature.  The well-formed
-		// check ensures that this transaction cannot pay any fee, and
-		// cannot have any other interesting fields, except for the compact
-		// cert payload.
-		if s.Txn.Sender == transactions.CompactCertSender && s.Txn.Type == protocol.CompactCertTx {
-			return nil
-		}
+	return
+}
+
+// Special case: special sender address can issue special transaction
+// types (compact cert txn) without any signature.  The well-formed
+// check ensures that this transaction cannot pay any fee, and
+// cannot have any other interesting fields, except for the compact
+// cert payload.
+func isCompactCertSpecialCase(s *transactions.SignedTxn) bool {
+	return s.Txn.Sender == transactions.CompactCertSender && s.Txn.Type == protocol.CompactCertTx
+}
 
+func TxnIsMissingSig(s *transactions.SignedTxn) bool {
+	numSigs, _, _, _ := identifySigs(s)
+	return numSigs == 0 && !isCompactCertSpecialCase(s)
+}
+
+func stxnVerifyCore(s *transactions.SignedTxn, txnIdx int, groupCtx *GroupContext, batchVerifier *crypto.BatchVerifier) error {
```

Comment:
## Logical equivalence proof

### Existing Implementation
- `numsigs == 0`
  - Yes: Is compact cert special case?
    - Yes: `return nil`
    - No: Error with "no signature"
  - No: Continue checking signatures

### New Implementation (Full)
- `numsigs == 0`
  - Yes: Is compact cert special case?
    - Yes: `numsigs == 0`
      - Yes: `return nil`
      - No: *This case cannot exist because we are in the Yes branch of the parent `numsigs == 0`*
    - No: Error with "no signature"
  - No: `numsigs == 0`
    - Yes: *This case cannot exist because we are in the No branch of the parent `numsigs == 0`*
    - No: Continue checking signatures

### New Implementation (Simplified)

After simplifying single-branch `numsigs == 0` branches, we find:

- `numsigs == 0`
  - Yes: Is compact cert special case?
    - Yes: `return nil`
    - No: Error with "no signature"
  - No: Continue checking signatures

Reply:
Is this work because you don't want to rely on testing the error that is returned? I figured you would just run the existing code and figure out why it bailed out (if it does).
```
	_, err = verify.TxnGroup(txgroup, hdr, nil)
	if err != nil && !errors.Is(err, verify.MissingSignatureError) {
		return err
	}
```
That just requires substituting in that specific error in the two cases that complain about missing signatures. (The normal one, and I think you want to allow callers to use a "delegated" logic sig in which the delegation signature is missing.)

Reply:
No, it's because I realized that `verify.TxnGroup` will short-circuit future well-formed checks when a signature was found to be missing.

This meant that it was possible for an invalid transaction group to succeed incorrectly. In fact, I discovered an instance of that behavior in the simulator tests where I was including approval and clear programs incorrectly when also passing an app index.

Reply:
I see, good point.

Unfortunately, a missing signature will also mean that pooled fees aren't properly checked.  I suppose that's not the worst thing.  I guess you can recheck them as well, though that check is already duplicated in the codebase, I think. It would be a shame to have it in three times.

Reply:
I believe I've fixed this with my latest commits. I'm not too pleased with the fix, but I couldn't think of a cleaner solution.

Reply:
I've managed to find a solution that does not modify this file, which I've committed in 139672aeae82f0eb80b86171c6d6b4ebcbd5d6ae.

Resolving this thread.